pcdsdevices/dg645.py

```diff
@@ -0,0 +1,447 @@
+"""
+Module for DG645 delay generator timing channel.
+"""
+from ophyd.device import Device, Component as Cpt
+from ophyd.signal import EpicsSignal, EpicsSignalRO
+from ophyd.utils import LimitError
+
+
+class DG645(object):
+    """
+    Class to access all channels of DG645 delay generator
+    """
+    def __init__(self, prefix, name='dg645', **kwargs):
+        self.chAB = DG645Channel(prefix, 'AB')
+        self.chCD = DG645Channel(prefix, 'CD')
+        self.chEF = DG645Channel(prefix, 'EF')
+        self.chGH = DG645Channel(prefix, 'GH')
+
+
+class DG645Channel(object):
+    """
+    Defines a particular channel with pulse output `XY` and delay outputs
+    `X` and `Y`. Set width, amplitude, polarity, reference and triggers 
+    for pulse. Limits can be passed as kwargs.
+
+    Parameters
+    ---------
+    prefix : ``str``
+        The PV base of the relevant DG, e.g. 'MEC:LAS:DDG:01:'
+    
+    channel : ``str``
+        The pulse channel on the DG, e.g. 'AB', 'CD', etc
+
+    refA : ``str``, optional
+        Leading edge zero-time reference, typically 'T0'
+
+    refB : ``str``, optional
+        Reference channel for following edge, typically rising edge channel 
+    
+    delay_limits : ``tuple``, optional
+
+    amp_limits : ``tuple``, optional
+    
+    Usage
+    ---------
+    DG645Channel('MEC:LAS:DDG:01:', 'AB', amp_limits=(0.0, 3.0) )
+    """
+    
+    def __init__(self, prefix, channel, refA='T0', refB=None, **kwargs):
+        self.ampCh   = pulseChannel(prefix, channel, **kwargs)
```

Comment:
I think we can restructure things to get `DG645Channel` to be a `Device` as well. It would have 3 components and make use of (the admittedly somewhat confusing) `FormattedComponent`. It'd be something along the lines of:

```python
from ophyd.device import FormattedComponent as FCpt

class DG645Channel(Device):
    ampCh = FCpt(pulseChannel, '{self._amp_prefix}')
    delayCh = FCpt(delayChannel, '{self._delay_prefix}')
    widthCh = FCpt(delayChannel, '{self._width_prefix}')

    def __init__(self, prefix, channel, refA="T0", refB=None, **kwargs):
        self._amp_prefix = f'{prefix}{channel[0].lower()}'
        self._delay_prefix = f'{prefix}{channel[0].lower()}'
        self._width_prefix = f'{prefix}{channel[1].lower()}'
        super().__init__(prefix, **kwargs)
```

This means that we can assemble the prefix dynamically for each of the components and still use the entire `Device` structure.

Reply:
Is there any way to pass 2 separate string arguments to the pulseChannel (or delayChannel) using FCpt? When I try the above syntax, it breaks the __init__() method in pulseChannel because it wants the channel argument separate from the prefix argument. Alternately, I could rewrite the pulse/delay classes to take only one positional argument, and recover the channel from the prefix. Either works
`TypeError: __init__() missing 1 required positional argument: 'channel'`

Reply:
You can add in non-dynamic keyword arguments to the `Cpt()` portion, but everything else will have to be set up after the call to `super().__init__` above.

That is to say, you could just use something like: `self.ampCh.channel = channel[0]`
If nothing complex needs to happen when `channel` is set, then that should be sufficient. Otherwise, you could always create a method such as `set_channel`  (or a property, similarly).

It doesn't seem like `'.channel'` is being used elsewhere, so that may not even be necessary.

(if you need help with further details on this, chat with @teddyrendahl or @ZLLentz. I'm headed out for vacation and plan to go GitHub-free for a while :) Good luck!)